doc/sailfish-webview-qmozcontext.cpp

```diff
@@ -0,0 +1,367 @@
+    set a custom profile location in case the intialization process is being
+    overridden.
+
+    \note If using a different location care is needed to select a path that's
+    acceptable from an application sandboxing point of view. The qstandard
+    application paths are safe for profile paths.
+
+    See \l {SailfishOS::WebEngine::initialize}{WebEngine::initialize} for more
+    info.
+
+    \sa {SailfishOS::WebEngine::initialize}{WebEngine::initialize}, runEmbedding, stopEmbedding
+*/
+void QMozContext::setProfile(const QString &profilePath);
+
+/*!
+    \internal
+    \brief Called automatically on WebView creation so that the engine can create views.
+
+    For internal use.
+
+    \sa setViewCreator
+*/
+void QMozContext::setViewCreator(QMozViewCreator *viewCreator);
+
+/*!
+    \internal
+    \brief Opens a new view that can contain webcontent.
+
+    For internal use.
+
+    \sa setViewCreator
+*/
+void QMozContext::createView();
+
+/*!
+    \brief This signal is emitted when a notification on an observed topic is received.
+
+    After an app has registered an interest in a particular topic by calling
+    \l addObserver or \l addObservers, it can then conect to this signal to be
+    notified about any component sending a notification on the topic.
+
+    The \a message parameter will be the topic in question, the \a data parameter
+    will contain data sent by the sender constructed from a JSON object. The
+    exact object structure is at the sender's discretion.
+
+    \sa addObserver, addObservers, removeObserver, removeObservers, notifyObservers
+*/
+void QMozContext::recvObserve(const QString message, const QVariant data);
+
+/*!
+    \brief This signal is emitted when the gecko EmbedLiteApp has been destroyed.
+
+    This signal is emitted when the EmbedLiteApp has been destroyed, just prior
+    to the WebEngine being deleted and the program exiting.
+*/
+void QMozContext::contextDestroyed();
+
+/*!
+    \brief This signal is emitted when when the last view has been destroyed.
+*/
+void QMozContext::lastViewDestroyed();
+
+/*!
+    \brief This signal is sent when the last registered window has been destroyed.
+
+    By default this is used to delay execution of \l stopEmbedding until after
+    the last window has closed.
+*/
+void QMozContext::lastWindowDestroyed();
+
+/*!
+    \brief Regsiter javascript chrome components to be loaded into the WebEngine.
+
+    The sailfish WebEngine has three main layers. The top layer is exposed as
+    QML components with a developer-facing interface. The lowest level is the
+    native gecko rendering engine.
+
+    Between the two sits a javascript chrome layer that can communicate with
+    either of the other layers using an observer/notification pattern. These
+    javascript components are considered privileged code and run with the
+    \l {https://developer.mozilla.org/en-US/docs/Mozilla/Gecko/Script_security#security_principals}{system principal}.
```

Comment:
This worries me slightly - is there a specific permission that the app has to request, in order to use this?  I don't like the idea of an application showing me a web view, which can read credentials/content from sites accessed via the webview... I mean, I guess a normal webview is loaded in-process anyway, so it's possible to read the contents of memory regardless, but still ...

Reply:
There's no permission. What you say is accurate and I hear your concern, but realistically I wonder if there's any way to avoid this. Even if this capability weren't available the app could display a fake local copy of a website, ask the user to enter their credentials, capture the credentials, then forward the user to the real website, all without the user knowing. As there's no "trusted TLS padlock" in the webview, the user wouldn't be able to tell the difference. I presume this is why Android doesn't allow OAuth to be performed inside apps any more (to discourage users from entering their credentials into anything except the system browser).